### PR TITLE
DSDRs

### PR DESCRIPTION
**Pull request by otfrom**

### CODE REVIEW

File: doc/dsdr/0010-placement-modelling.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+H records: historically closed cases. We sample probable birthdays for each of these using the birth month as a key constraint, but also the date of entry (so joiners can't be less than zero) and the leave date (so that leavers can't be over 18).
+O records: historically open cases. We sample probable birthdays for these as above but using the extract date instead of leave date (since none yet exists).
+C records: candidate periods. These are the open periods with resampled birthdays, and we create multiple birthdays for each candidate by resampling n times, specified by the `:candidate-variations` config variable.
+P records consisting of: a) C records which are completed according to the baseline period completer algorithm (described below).
+  b) C records which are completed according to the 'age out' period completer algorithm.
+S records consisting of: a) C records which are truncated to an arbitrary point in their past and then completed with the baseline period completer algorithm.
+  b) C records which are truncated to an arbitrary point in their past and then completed with the 'age out' period completer algorithm.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:28]: This isn't rendering as a list in markdown

File: doc/dsdr/0010-placement-modelling.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+Since the model needs to close open cases (as well as generate brand-new projected cases), we must make sure that the `generate-candidates!` step knows when our projection starts because this determines which cases are open at that point in time. This date may not be the day after the data was extracted, since it's valuable to 'back-cast' the projection from earlier points in time for cross-validation. As a result, the generate candidates output is specific to a particular projection start date, and can't be re-used if a different projection start date is subsequently desired: the `generate-candidates step!` must be run again with the new projection start date.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:19]: second `generate-candidates step!` should be `generate-candidates` step!

File: doc/dsdr/0010-placement-modelling.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
...ave a way of assigning a plausible placement to each case at each point in time. The placements and the transitions between them should exhibit similar characteristics to those that have been observed historically according to the projected child's age and case history. Several strategies have been attempted to meet this requirement, most of which have involved creating future pathways as a composite of shorter building blocks sampled from the historic data. These methods have tended not to model total durations in care very well. The probability of a child leaving care is based on factors including their current age, ther placement, and their prior duration in care. This final factor is particularly difficult to model because it's non-linear: for most ages the probability of leaving is bimodal, with one mode falling within the first few years, and the second peak occuring prior to the 18th birthday with only a small probability of leaving in between. The number of projected children in care is extremely sensitive to the child's duration in care because small differences in the average duration apply to all current and future children, resulting in something like a compound effect.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:5]: ther placement/their placement

File: doc/dsdr/0010-placement-modelling.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+![Segmentation](https://raw.githubusercontent.com/MastodonC/witan.cic/dsdrs/doc/img/segmentation.png)
+
+The constraint that all segments except terminal segments are sourced by segmenting existing periods into sequential 6-year chunks implies a potential issue completing an open case of duration 9 years: would we match to a segment captured starting after 6 years or after 12 years? Because neither may be a good match, we implement an 'offset' algorithm for splicing segments other than at a 6-year boundary. This algorithm starts by calculating the offset through the latest 6-year window of the end of the open case. For 9 years this might be 3 * 365 = 1095 days, because the 9th year has a 3 year offset into the second segment. We also create offset versions of all segments, skipping the first n days to create tail portions. In order to complete the matching efficiently, we pre-calculate all offset versions of all segments in advance. An offset version of a terminal segment is still a terminal segment.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:40]: I'm still not getting why 9 years is a magic year.

File: doc/dsdr/0010-placement-modelling.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+*Splitting periods into segments*
+
+Each period is split into consecutive segments of 6 years. For closed cases, the partial segment at the end is retained and marked as a terminal segment. For open cases, the final segment at the end is discarded. If segments start occur within the training range (specified as n years before projection start), they're marked as `:in-filter?`. This capability allows for the possibility of filtering candidate segments to within a particular date range, although this functionality is not currently used.
+
+![Segmentation](https://raw.githubusercontent.com/MastodonC/witan.cic/dsdrs/doc/img/segmentation.png)
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:38]: I *think* you can do these as relative links

File: doc/dsdr/0010-placement-modelling.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
...ching of certain fields ensures that segments being chained are compatible, whereas fuzzy matching ensures that the algorithm is non-deterministic and introduces useful variety into the generated ccandidates. Fields which are matched exactly are the placement at the join, the offset (measured in days) from a round 6 years and whether or not a 'terminal segment' is desired. Fuzzy matches are the join age in days and the number of days in care. Fuzzy matching is implemented by adding noise to the match values and then finding the closest matching segment using the Euclidean distance. Because match values of join age and days in care have well-defined min and max values (0 days and the durations corresponding to the 18th birthday respectively), binomial (instead of guassian) noise is added using `cic.model/jitter-binomial`. The binomial distribution of noise has several useful properties: it ensures that values are constrained within the min and max whilst remaining balanced around the mean, and has smaller variance at the extremes (close to zero and 18 years old) where timings are more sensitive and increasing variance in the middle of the range where timings are much less sensitive.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:50]: Useful to have a pointer to the matching function

File: doc/dsdr/0010-placement-modelling.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
...nd red I could employ rejection sampling after each sample. Every time I draw a blue marble from the bag I could sample a random number between 0 and 1. If the number is greater than 10/5, or 0.5 (the rejection threshold), I would log a blue sample and return the marble. If it is less than 0.5 I would return the marble without logging the sample and repeat the process. For red marbles the rejection threshold would be 0. If I repeated this process lots of times, the accepted samples would show a distribution equivalent to if I had an even number of blue and red marbles in the bag - a uniform distribution. Careful calibration of the rejection thresholds (0.5 and 0 in this example) for each sampled value can theoretically turn any candidate distribution into any other target distribution, provided the candidate distribution has some probability density everywhere the target distribution has probability density. This is because rejection sampling can only reject proposed samples, it can't create samples if they're not generated in thef irst place. This is equivalent to saying that we can't create a uniform reb-blue-green marble distribution if there aren't any green marbles in the bag.
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:62]: Is there a good description of rejection sampling we could link to? A formal and/or an informal one?

Comment (otfrom) [reply doc/dsdr/0010-placement-modelling.md:62]: tho, this is a good explanation and the first time I properly understand how the rejection sampling workd.

File: doc/dsdr/0010-placement-modelling.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+**Age out model**
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:70]: I like the explanation in this whole section

File: doc/dsdr/0010-placement-modelling.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+Supporting this extra model requires an additional model input: the probability of exceeding 17 years of age in care by admission age. This is close to 0 for zero-age joiners, and exactly 1 for age 17 joiners. It's contained in the file `age-out-proportions.csv`.
+
+**Candidates files**
+
+Generate candidates process outputs 4 files: `simulated-candidates.csv`, `projected-candidates.csv`, `projected-age-out-candidates.csv`, `simulated-age-out-candidates.csv`.
+
+The `simulated-*` candidates files are organised by admission age and are used for generating new joiners, the `projected-*` files are indexed by period id and are used for closing open cases.
+The `*-age-out-candidates` files contain the results of attempting to generate closed cases which persist until 18, whereas the other files will generally not contain age out candidates. This is not _always_ the case: we will have been unable to generate projected-age-out candidates for some children (if there is nothing in the source data which allows us to chain a sequence which ends at 18 for that child). Likewise there will be age 17 joiners in the non-age-out files who age out by definition.
+
+**Projection**
```
Comment (otfrom) [inline doc/dsdr/0010-placement-modelling.md:83]: This section could use a couple of flow charts, but is still pretty understandable.

**Review by otfrom — COMMENTED** [review]
A few tweaks pls. I understand a lot more now that I did before.

File: doc/dsdr/0010-placement-modelling.md (hunk 5 above)
Comment (henrygarner) [reply doc/dsdr/0010-placement-modelling.md]: So you can! 👍

File: doc/dsdr/0010-placement-modelling.md (hunk 6 above)
Comment (henrygarner) [reply doc/dsdr/0010-placement-modelling.md]: I'm not sure which bit of the text you wanted the matching function for, but I've added references to `cic.model/markov-period` which does most of the work, and `cic.model/get-matched-segment` which is how the closest match is identified using the Euclidean distance.

**Pull request merged**